libflux/src/core/semantic/types.rs

```diff
@@ -663,6 +671,57 @@ impl Array {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct Dictionary {
+    pub key: MonoType,
+    pub val: MonoType,
+}
+
+impl fmt::Display for Dictionary {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}:{}]", self.key, self.val)
+    }
+}
+
+impl Substitutable for Dictionary {
+    fn apply(self, sub: &Substitution) -> Self {
+        Dictionary {
+            key: self.key.apply(sub),
+            val: self.val.apply(sub),
+        }
+    }
+    fn free_vars(&self) -> Vec<Tvar> {
+        union(self.key.free_vars(), self.val.free_vars())
+    }
+}
+
+impl MaxTvar for Dictionary {
+    fn max_tvar(&self) -> Tvar {
+        vec![self.key.max_tvar(), self.val.max_tvar()].max_tvar()
+    }
+}
+
+impl Dictionary {
+    fn unify(
```

Comment:
I'm unclear on what `unify`ing means in this context. Could you elaborate?

Reply:
In the type system, unification is the process of resolving equality constraints, the result of which is a substitution which maps type variables to the monotypes they represent. For example, unification of two primitive types only succeeds (producing an empty substitution) if both types are the same
```
int = int
string = string
int != string
```

Dictionary types are composite types, so two dictionary types unify if and only if both of their components unify
```
[k0:v0] = [k1:v1] iff k0 = k1 and v0 = v1
```

Reply:
Awesome. That makes sense. Thanks for the clarification :+1: